fs/unpacker.go

```diff
@@ -46,56 +47,184 @@ type Archive interface {
 }
 
 type layerArchive struct {
-	verifier         *layerVerifier
+	compressed       *asyncVerifier
+	uncompressed     *asyncVerifier
 	decompressStream socicompression.DecompressStream
+	bufferPool       *bufferPool
 }
 
-type layerVerifier struct {
-	compressed   digest.Verifier
-	uncompressed digest.Verifier
+type asyncVerifier struct {
+	v       digest.Verifier
+	waitCh  chan struct{}
+	started bool
 }
 
-func NewLayerArchive(compressedVerifier, uncompressedVerifier digest.Verifier, decompressStream socicompression.DecompressStream) Archive {
+func newAsyncVerifier(v digest.Verifier) *asyncVerifier {
+	return &asyncVerifier{
+		v:       v,
+		waitCh:  make(chan struct{}, 1),
+		started: false,
+	}
+}
+
+func (av *asyncVerifier) AsyncVerify(reader io.ReadCloser) {
+	if av == nil || av.v == nil {
+		return
+	}
+	sig := make(chan struct{}, 1)
+	go func() {
+		close(sig)
+		io.Copy(av.v, reader)
+		close(av.waitCh)
+		reader.Close()
+	}()
+	<-sig
+	av.started = true
+}
+
+func (av *asyncVerifier) Verified(ctx context.Context) bool {
+	if av == nil || av.v == nil || !av.started {
+		return true
+	}
+	select {
+	case <-ctx.Done():
+		return false
+	case <-av.waitCh:
+		return av.v.Verified()
+	}
+}
+
+type bufferPool struct {
+	pool *sync.Pool
+}
+
+func newbufferPool(size int64) *bufferPool {
+	pool := &sync.Pool{
+		New: func() any {
+			buffer := make([]byte, size)
+			return &buffer
+		},
+	}
+	return &bufferPool{
+		pool: pool,
+	}
+}
+
+func (p *bufferPool) Get() *[]byte {
+	buf := p.pool.Get().(*[]byte)
+	return buf
+}
+
+func (p *bufferPool) Put(buffer *[]byte) {
+	p.pool.Put(buffer)
+}
+
+func NewLayerArchive(compressedVerifier, uncompressedVerifier *asyncVerifier, decompressStream socicompression.DecompressStream, bufPool *bufferPool) Archive {
 	// If no layer decompress stream was provided, then use containerd's decompress stream implementation.
 	if decompressStream == nil {
 		decompressStream = compression.DecompressStream
 	}
+	if bufPool == nil {
+		bufPool = newbufferPool(64 * 1024)
+	}
 	return &layerArchive{
-		verifier: &layerVerifier{
-			compressed:   compressedVerifier,
-			uncompressed: uncompressedVerifier,
-		},
+		compressed:       compressedVerifier,
+		uncompressed:     uncompressedVerifier,
 		decompressStream: decompressStream,
+		bufferPool:       bufPool,
 	}
 }
 
-func (la *layerArchive) Apply(ctx context.Context, root string, r io.Reader, opts ...archive.ApplyOpt) (int64, error) {
-	// Decompress first, then apply.
-	if la.verifier.compressed != nil {
-		r = io.TeeReader(r, la.verifier.compressed)
+func AsyncTeeReader(r io.Reader, w io.WriteCloser, bp *bufferPool) io.Reader {
```

Comment:
Did you consider an `AsyncWriter` + normal io.TeeReader for this? I'm not sure if there's a use-case for an async writer outside of a TeeReader, but maybe that would be a smaller abstraction? What do you think?

Reply:
There is already bufio in golang but the challenge with it is that golang grows the buffer cap as needed opposed to limiting the cap to a fixed size. I considered to implement a AsyncWriter with circular buffers, but I don't really see other use cases other than this. Also IMHO the AsyncTeeReader implementation is also simple enough. So I prefer keeping it as current form now.